zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/InterpreterSetting.java

```diff
@@ -713,16 +717,138 @@ RemoteInterpreterProcess createInterpreterProcess() {
+  private boolean isSparkConf(String key, String value) {
+    return !StringUtils.isEmpty(key) && key.startsWith("spark.") && !StringUtils.isEmpty(value);
+  }
+
   private Map<String, String> getEnvFromInterpreterProperty(Properties property) {
-    Map<String, String> env = new HashMap<>();
-    for (Object key : property.keySet()) {
-      if (RemoteInterpreterUtils.isEnvString((String) key)) {
-        env.put((String) key, property.getProperty((String) key));
+    Map<String, String> env = new HashMap<String, String>();
+    Properties sparkProperties = new Properties();
+    String sparkMaster = null;
+    for (String key : property.stringPropertyNames()) {
+      if (RemoteInterpreterUtils.isEnvString(key)) {
+        env.put(key, property.getProperty(key));
+      }
+      if (key.equals("master")) {
+        sparkMaster = property.getProperty("master");
       }
+      if (isSparkConf(key, property.getProperty(key))) {
+        sparkProperties.setProperty(key, toShellFormat(property.getProperty(key)));
+        if (key.equals("spark.master")) {
+          if (sparkMaster == null) {
+            sparkMaster = property.getProperty(key);
+          }
+        }
+      }
+    }
+
+    setupPropertiesForPySpark(sparkProperties);
+    setupPropertiesForSparkR(sparkProperties);
+    if (sparkMaster != null && sparkMaster.equals("yarn-cluster")) {
+      env.put("SPARK_YARN_CLUSTER", "true");
+    }
+
+    StringBuilder sparkConfBuilder = new StringBuilder();
+    if (sparkMaster != null) {
+      sparkConfBuilder.append(" --master " + sparkMaster);
+    }
+    sparkConfBuilder.append(" --files " + conf.getConfDir() + "/log4j_yarn_cluster.properties");
+    for (String name : sparkProperties.stringPropertyNames()) {
+      sparkConfBuilder.append(" --conf " + name + "=" + sparkProperties.getProperty(name));
     }
+
+    env.put("ZEPPELIN_SPARK_CONF", sparkConfBuilder.toString());
     return env;
   }
 
+  private void setupPropertiesForPySpark(Properties sparkProperties) {
+    Object pysparkBaseProperty =
+        new DefaultInterpreterProperty("SPARK_HOME", null, null).getValue();
+    String pysparkBasePath = pysparkBaseProperty != null ? pysparkBaseProperty.toString() : null;
+    File pysparkPath;
+    if (null == pysparkBasePath) {
+      pysparkBasePath =
+          new DefaultInterpreterProperty("ZEPPELIN_HOME", "zeppelin.home", "../")
+              .getValue().toString();
+      pysparkPath = new File(pysparkBasePath,
+          "interpreter" + File.separator + "spark" + File.separator + "pyspark");
+    } else {
+      pysparkPath = new File(pysparkBasePath,
+          "python" + File.separator + "lib");
+    }
+
+    //Only one of py4j-0.9-src.zip and py4j-0.8.2.1-src.zip should exist
+    //TODO(zjffdu), this is not maintainable when new version is added.
+    String[] pythonLibs = new String[]{"pyspark.zip", "py4j-0.9-src.zip", "py4j-0.8.2.1-src.zip",
+      "py4j-0.10.1-src.zip", "py4j-0.10.3-src.zip", "py4j-0.10.4-src.zip"};
```

Comment:
we really need to fix this....

Reply:
Yes, I wouldn't have thought it used for a long time...

Reply:
Sure, this is on my plan for the next step.